File: backend/services/exam_service.py
from io import BytesIO
from urllib.parse import quote
from flask import Response
from pydantic import BaseModel as _BaseModel
from extensions import db
from repositories.exam_repository import ExamRepository, StudentAnswerRepository
from repositories.class_repository import ClassExamRepository
from errors import NotFoundError, ConflictError
from services.pdf_service import build_exam_pdf
from config import init_chat_model
import openpyxl


class _Feedback(_BaseModel):
    feedback: str

class _Insight(_BaseModel):
    title: str
    detail: str

class _StatsAnalysis(_BaseModel):
    insights: list[_Insight]

exam_repo    = ExamRepository()
sa_repo      = StudentAnswerRepository()
ce_repo      = ClassExamRepository()


def get_exam(exam_id: int, student_id: int = None) -> dict:
    exam = exam_repo.find_by_id(exam_id)
    if not exam:
        raise NotFoundError(" Bài tập không tồn tại")
    time_limit = None
    if student_id:
        row = ce_repo.find_for_student(student_id, exam_id)
        time_limit = int(row["time_limit"]) if row else 20 * 60
    return {
        "name":        exam.name,
        "description": exam.description,
        "dateCreated": str(exam.date_created) if exam.date_created else None,
        "timeLimit":   time_limit,
        "questions": [
            {
                "questionId":      q.id,
                "questionContent": q.content,
                "explanation":     q.explanation,
                "answers": [
                    {"answerId": a.id, "content": a.content, "isCorrected": a.is_correct}
                    for a in q.answers
                ],
            }
            for q in exam.questions
        ],
    }


def get_assignments(exam_id: int, teacher_id: int) -> list:
    rows = ce_repo.find_by_exam_and_teacher(exam_id, teacher_id)
    for r in rows:
        r["assigned"]    = bool(r["assigned"])
        r["deadline"]    = str(r["deadline"])    if r["deadline"]    else None
        r["assigned_at"] = str(r["assigned_at"]) if r["assigned_at"] else None
        r["open_time"]   = str(r["open_time"])   if r["open_time"]   else None
        r["time_limit"]  = int(r["time_limit"])  if r["time_limit"] is not None else None
    return rows


def create_exam(lesson_id: int, name: str, description: str, questions: list) -> int:
    exam = exam_repo.create(lesson_id, name, description, questions)
    db.session.commit()
    return exam.id


def update_exam(exam_id: int, lesson_id: int, name: str, description: str, questions: list) -> None:
    exam = exam_repo.find_by_id(exam_id)
    if not exam or exam.lesson_id != lesson_id:
        raise NotFoundError(" Bài tập không tồn tại")
    exam_repo.update(exam, name, description, questions)
    db.session.commit()


def delete_exam(exam_id: int) -> None:
    exam = exam_repo.find_by_id(exam_id)
    if not exam:
        raise NotFoundError(" Bài tập không tồn tại")
    exam_repo.delete(exam)
    db.session.commit()


def clone_exam(exam_id: int) -> int:
    exam = exam_repo.find_by_id(exam_id)
    if not exam:
        raise NotFoundError(" Bài tập không tồn tại")
    new_exam = exam_repo.clone(exam)
    db.session.commit()
    return new_exam.id


def submit_exam(exam_id: int, student_id: int, answers: list, time_spent: int) -> dict:
    if sa_repo.has_submitted(student_id, exam_id):
        raise ConflictError("Bạn đã nộp bài rồi!")
    result = sa_repo.submit(student_id, exam_id, answers, time_spent)
    db.session.commit()
    return result


def get_result(exam_id: int, student_id: int) -> dict:
    data = exam_repo.get_exam_result(exam_id, student_id)
    if not data:
        raise NotFoundError(" Bài tập không tồn tại")
    return data


def get_stats(exam_id: int) -> dict:
    exam = exam_repo.find_by_id(exam_id)
    if not exam:
        raise NotFoundError(" Bài tập không tồn tại")
    raw = exam_repo.get_stats(exam_id)
    scores         = raw["scores"]
    total_students = raw["total_students"]
    question_rows  = raw["question_rows"]

    dist = {"0-4": 0, "4-6": 0, "6-8": 0, "8-10": 0}
    for s in scores:
        if s < 4:   dist["0-4"] += 1
        elif s < 6: dist["4-6"] += 1
        elif s < 8: dist["6-8"] += 1
        else:       dist["8-10"] += 1

    questions_stats = []
    for qr in question_rows:
        ta = int(qr["totalAnswered"] or 0)
        cc = int(qr["correctCount"]  or 0)
        questions_stats.append({
            "questionId":    qr["questionId"],
            "content":       qr["content"],
            "totalAnswered": ta,
            "correctCount":  cc,
            "correctRate":   round(cc / ta * 100, 1) if ta > 0 else 0,
        })

    return {
        "examName":          exam.name,
        "totalStudents":     total_students,
        "completedStudents": len(scores),
        "avgScore":          round(sum(scores) / len(scores), 1) if scores else 0,
        "distribution":      dist,
        "questions":         questions_stats,
    }


def generate_ai_feedback(exam_id: int, student_id: int) -> str:
    data  = get_result(exam_id, student_id)
    wrong = [q for q in data.get("questions", []) if not q.get("isCorrect")]
    if not wrong:
        return "Xuất sắc! Em đã trả lời đúng tất cả các câu hỏi. Hãy tiếp tục phát huy nhé!"
    lines = []
    for q in wrong:
        correct = next((a["content"] for a in q.get("answers", []) if a.get("isCorrected") == 1), "")
        chosen  = next((a["content"] for a in q.get("answers", []) if a.get("isSelected") and a.get("isCorrected") != 1), "không rõ")
        lines.append(f'- "{q["questionContent"]}"\n  Đáp án đúng: {correct} | Em chọn: {chosen}')
    prompt = f"""Bạn là giáo viên Toán lớp 3 đang nhận xét bài làm của học sinh.
Bài kiểm tra: {data.get("examName", "Toán lớp 3")}
Kết quả: {data.get("score", 0)}/10 — đúng {data.get("correct", 0)}/{data.get("total", 0)} câu.

Các câu trả lời sai:
{chr(10).join(lines)}

Viết 1 đoạn nhận xét (2-4 câu) bằng tiếng Việt, thân thiện dành cho học sinh lớp 3: chỉ ra lỗi sai cụ thể và cách khắc phục, kết bằng lời khích lệ.
"""
    result = init_chat_model().generate(prompt, response_model=_Feedback)
    return result.feedback


def analyze_exam_stats(exam_id: int) -> list:
    stats = get_stats(exam_id)
    if not stats.get("completedStudents"):
        return []
    qs = stats.get("questions", [])
    hard = [q for q in qs if q.get("totalAnswered", 0) > 0 and q.get("correctRate", 100) < 50]
    if not hard:
        hard = sorted(qs, key=lambda q: q.get("correctRate", 100))[:3]
    hard_summary = "\n".join(
        f'- Câu {i+1}: "{q["content"][:80]}" — tỉ lệ đúng {q["correctRate"]}% ({q["correctCount"]}/{q["totalAnswered"]} HS)'
        for i, q in enumerate(hard)
    )
    dist = stats.get("distribution", {})
    prompt = f"""Bạn là chuyên gia giáo dục Toán lớp 3. Phân tích kết quả bài kiểm tra:
Tên bài: {stats.get("examName", "")}
Học sinh nộp bài: {stats.get("completedStudents", 0)}/{stats.get("totalStudents", 0)}
Điểm trung bình: {stats.get("avgScore", 0)}/10
Phân bố: 0-4: {dist.get("0-4",0)} HS, 4-6: {dist.get("4-6",0)} HS, 6-8: {dist.get("6-8",0)} HS, 8-10: {dist.get("8-10",0)} HS.

Câu hỏi học sinh làm sai nhiều nhất:
{hard_summary}

Đưa ra CHÍNH XÁC 3 nhận xét và lời khuyên thực tế cho giáo viên Toán lớp 3 dựa trên dữ liệu trên, trả lời bằng tiếng việt.
"""
    result = init_chat_model().generate(prompt, response_model=_StatsAnalysis)
    return [i.model_dump() for i in result.insights]


def save_comment(exam_id: int, student_id: int, teacher_id: int, comment: str) -> None:
    exam_repo.upsert_comment(exam_id, student_id, teacher_id, comment)
    db.session.commit()


def export_pdf(exam_id: int, variants_count: int, duration: int) -> Response:
    exam = exam_repo.find_by_id(exam_id)
    if not exam:
        raise NotFoundError(" Bài tập không tồn tại")
    questions = [
        {
            "question_id": q.id,
            "content":     q.content,
            "answers": [
                {"answer_id": a.id, "content": a.content, "is_correct": bool(a.is_correct)}
                for a in q.answers
            ],
        }
        for q in exam.questions
    ]
    pdf_bytes = build_exam_pdf(exam.name, duration, questions, variants_count)
    safe_name = quote(exam.name)
    return Response(
        pdf_bytes,
        mimetype="application/pdf",
        headers={"Content-Disposition": f"attachment; filename*=UTF-8''{safe_name}.pdf"},
    )


def export_results(exam_id: int) -> Response:
    exam = exam_repo.find_by_id(exam_id)
    if not exam:
        raise NotFoundError(" Bài tập không tồn tại")
    rows = exam_repo.get_export_data(exam_id)

    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "Kết quả thi"
    ws.append(["STT", "Họ tên", "Username", "Điểm", "Số câu đúng", "Tổng câu", "Thời gian (giây)", "Thời gian nộp"])
    for i, row in enumerate(rows, 1):
        ws.append([
            i, row["full_name"], row["username"],
            float(row["score"] or 0), int(row["correct_count"] or 0),
            int(row["total_questions"] or 0), int(row["time_spent"] or 0),
            str(row["submitted_at"]) if row["submitted_at"] else "",
        ])

    buf = BytesIO()
    wb.save(buf)
    buf.seek(0)
    safe_name = quote(exam.name)
    return Response(
        buf.getvalue(),
        mimetype="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename*=UTF-8''{safe_name}.xlsx"},
    )

File: backend/src/services/exam.service.js
import * as examRepo from '../repositories/exam.repository.js';
import * as classRepo from '../repositories/class.repository.js';
import { NotFoundError, ConflictError, ForbiddenError } from '../utils/error.utils.js';
import { generateJSON } from '../utils/ai.js';
import { buildExamPdf } from './pdf.service.js';
import { formatDate, formatDateTime } from '../utils/date.utils.js';

export const getExam = async (examId, studentId = null) => {
  const exam = await examRepo.findWithQuestions(examId);
  if (!exam) throw new NotFoundError('Bài tập không tồn tại');
  let timeLimit = null;
  if (studentId) {
    const row = await classRepo.findClassExamForStudent(studentId, examId);
    timeLimit = row ? parseInt(row.time_limit) : 20 * 60;
  }
  return {
    name: exam.name,
    description: exam.description,
    dateCreated: formatDate(exam.date_created),
    timeLimit,
    questions: exam.questions.map(q => ({
      questionId: q.id,
      questionContent: q.content,
      explanation: q.explanation,
      answers: q.answers.map(a => ({
        answerId: a.id,
        content: a.content,
        isCorrected: a.is_correct,
      })),
    })),
  };
};

export const getAssignments = async (examId, teacherId) => {
  const rows = await classRepo.findClassExamsByExamAndTeacher(examId, teacherId);
  return rows.map(r => ({
    ...r,
    assigned: Boolean(r.assigned),
    deadline: formatDateTime(r.deadline),
    assigned_at: formatDateTime(r.assigned_at),
    open_time: formatDateTime(r.open_time),
    time_limit: r.time_limit != null ? parseInt(r.time_limit) : null,
  }));
};

export const createExam = async (lessonId, name, description, questions) => {
  return examRepo.createExam(lessonId, name, description, questions);
};

export const updateExam = async (examId, lessonId, name, description, questions) => {
  const exam = await examRepo.findById(examId);
  if (!exam || exam.lesson_id !== lessonId) throw new NotFoundError('Bài tập không tồn tại');
  await examRepo.updateExam(examId, name, description, questions);
};

export const deleteExam = async (examId) => {
  const exam = await examRepo.findById(examId);
  if (!exam) throw new NotFoundError('Bài tập không tồn tại');
  await examRepo.deleteExam(examId);
};

export const cloneExam = async (examId) => {
  const exam = await examRepo.findById(examId);
  if (!exam) throw new NotFoundError('Bài tập không tồn tại');
  return examRepo.cloneExam(examId);
};

export const submitExam = async (examId, studentId, answers, timeSpent) => {
  if (await examRepo.hasSubmitted(studentId, examId)) throw new ConflictError('Bạn đã nộp bài rồi!');
  return examRepo.submitExam(studentId, examId, answers, timeSpent);
};

export const getResult = async (examId, studentId) => {
  const data = await examRepo.getExamResult(examId, studentId);
  if (!data) throw new NotFoundError('Bài tập không tồn tại');
  return data;
};

export const getStats = async (examId) => {
  const exam = await examRepo.findById(examId);
  if (!exam) throw new NotFoundError('Bài tập không tồn tại');
  const raw = await examRepo.getStats(examId);
  const { scores, total_students, question_rows } = raw;

  const dist = { '0-4': 0, '4-6': 0, '6-8': 0, '8-10': 0 };
  for (const s of scores) {
    if (s < 4)      dist['0-4']++;
    else if (s < 6) dist['4-6']++;
    else if (s < 8) dist['6-8']++;
    else            dist['8-10']++;
  }

  const questionStats = question_rows.map(qr => {
    const ta = parseInt(qr.totalAnswered || 0);
    const cc = parseInt(qr.correctCount || 0);
    return {
      questionId: qr.questionId,
      content: qr.content,
      totalAnswered: ta,
      correctCount: cc,
      correctRate: ta > 0 ? parseFloat((cc / ta * 100).toFixed(1)) : 0,
    };
  });

  return {
    examName: exam.name,
    totalStudents: total_students,
    completedStudents: scores.length,
    avgScore: scores.length ? parseFloat((scores.reduce((a, b) => a + b, 0) / scores.length).toFixed(1)) : 0,
    distribution: dist,
    questions: questionStats,
  };
};

export const generateAiFeedback = async (examId, studentId) => {
  const data = await getResult(examId, studentId);
  const wrong = (data.questions || []).filter(q => !q.isCorrect);
  if (!wrong.length) return 'Xuất sắc! Em đã trả lời đúng tất cả các câu hỏi. Hãy tiếp tục phát huy nhé!';

  const lines = wrong.map(q => {
    const correct = (q.answers || []).find(a => a.isCorrected == 1)?.content || '';
    const chosen = (q.answers || []).find(a => a.isSelected && a.isCorrected != 1)?.content || 'không rõ';
    return `- "${q.questionContent}"\n  Đáp án đúng: ${correct} | Em chọn: ${chosen}`;
  });

  const prompt = `Bạn là giáo viên Toán lớp 3 đang nhận xét bài làm của học sinh.
Bài kiểm tra: ${data.examName || 'Toán lớp 3'}
Kết quả: ${data.score}/10 — đúng ${data.correct}/${data.total} câu.

Các câu trả lời sai:
${lines.join('\n')}

Viết 1 đoạn nhận xét (2-4 câu) bằng tiếng Việt, thân thiện dành cho học sinh lớp 3.
Trả về JSON: {"feedback": "..."}`;

  const result = await generateJSON(prompt);
  return result.feedback || '';
};

export const analyzeExamStats = async (examId) => {
  const stats = await getStats(examId);
  if (!stats.completedStudents) return [];

  const qs = stats.questions || [];
  let hard = qs.filter(q => q.totalAnswered > 0 && q.correctRate < 50);
  if (!hard.length) hard = [...qs].sort((a, b) => a.correctRate - b.correctRate).slice(0, 3);

  const hardSummary = hard.map((q, i) =>
    `- Câu ${i + 1}: "${q.content.slice(0, 80)}" — tỉ lệ đúng ${q.correctRate}% (${q.correctCount}/${q.totalAnswered} HS)`
  ).join('\n');

  const dist = stats.distribution || {};
  const prompt = `Bạn là chuyên gia giáo dục Toán lớp 3. Phân tích kết quả bài kiểm tra:
Tên bài: ${stats.examName}
Học sinh nộp bài: ${stats.completedStudents}/${stats.totalStudents}
Điểm trung bình: ${stats.avgScore}/10
Phân bố: 0-4: ${dist['0-4'] || 0} HS, 4-6: ${dist['4-6'] || 0} HS, 6-8: ${dist['6-8'] || 0} HS, 8-10: ${dist['8-10'] || 0} HS.

Câu hỏi học sinh làm sai nhiều nhất:
${hardSummary}

Đưa ra CHÍNH XÁC 3 nhận xét và lời khuyên thực tế cho giáo viên Toán lớp 3.
Trả về JSON: {"insights": [{"title": "...", "detail": "..."}, ...]}`;

  const result = await generateJSON(prompt);
  return result.insights || [];
};

export const saveComment = async (examId, studentId, teacherId, comment) => {
  await examRepo.upsertComment(examId, studentId, teacherId, comment);
};

export const exportPdf = async (examId, variantsCount, duration) => {
  const exam = await examRepo.findWithQuestions(examId);
  if (!exam) throw new NotFoundError('Bài tập không tồn tại');
  const questions = exam.questions.map(q => ({
    question_id: q.id,
    content: q.content,
    answers: q.answers.map(a => ({
      answer_id: a.id,
      content: a.content,
      is_correct: Boolean(a.is_correct),
    })),
  }));
  return buildExamPdf(exam.name, duration, questions, variantsCount);
};

export const exportResults = async (examId) => {
  const exam = await examRepo.findById(examId);
  if (!exam) throw new NotFoundError('Bài tập không tồn tại');
  const rows = await examRepo.getExportData(examId);
  return { examName: exam.name, rows };
};
